src/main/java/org/opensearch/security/support/StreamableRegistry.java

```diff
@@ -0,0 +1,138 @@
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import org.opensearch.OpenSearchException;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+
+import static org.opensearch.security.support.SafeSerializationUtils.prohibitUnsafeClasses;
+
+/**
+ * Registry for any class that does NOT implement the <code>Writeable</code> interface
+ * and needs to be serialized over the wire. Supports registration of writer and reader via <code>registerStreamable</code>
+ * for such classes and provides methods <code>writeTo</code> and <code>readFrom</code> for objects of such registered classes.
+ * <br/>
+ * Methods are protected and intended to be accessed from only within the package. (mostly by <code>Base64Helper</code>)
+ */
+public class StreamableRegistry {
+
+    private static final StreamableRegistry INSTANCE = new StreamableRegistry();
+    public final BiMap<Class<?>, Integer> classToIdMap = HashBiMap.create();
+    private final Map<Integer, Entry> idToEntryMap = new HashMap<>();
+
+    private StreamableRegistry() {
+        registerAllStreamables();
+    }
+
+    private static class Entry {
+        Writeable.Writer<Object> writer;
+        Writeable.Reader<Object> reader;
+
+        Entry(Writeable.Writer<Object> writer, Writeable.Reader<Object> reader) {
+            this.writer = writer;
+            this.reader = reader;
+        }
+    }
+
+    private Writeable.Writer<Object> getWriter(Class<?> clazz) {
+        if (!classToIdMap.containsKey(clazz)) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return idToEntryMap.get(classToIdMap.get(clazz)).writer;
+    }
+
+    private Writeable.Reader<Object> getReader(int id) {
+        if (!idToEntryMap.containsKey(id)) {
+            throw new OpenSearchException(String.format("No reader registered for id %s", id));
+        }
+        return idToEntryMap.get(id).reader;
+    }
+
+    private int getId(Class<?> clazz) {
+        if (!classToIdMap.containsKey(clazz)) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return classToIdMap.get(clazz);
+    }
+
+    protected boolean isStreamable(Class<?> clazz) {
+        return classToIdMap.containsKey(clazz);
+    }
+
+    protected void writeTo(StreamOutput out, Object object) throws IOException {
+        prohibitUnsafeClasses(object.getClass());
+        out.writeByte((byte) getId(object.getClass()));
+        getWriter(object.getClass()).write(out, object);
+    }
+
+    protected Object readFrom(StreamInput in) throws IOException {
+        int id = in.readByte();
+        prohibitUnsafeClasses(classToIdMap.inverse().get(id));
```

Comment:
Can be removed based on comments in Base64CustomHelper

Reply:
Done